File: Structural Pattern/Adapter/AdapterPatternExample.java
// package com.github.nikson.dpf

/*
	Adapter: it a bridge between two incompatible interface 
	Two types: Class Adapter, Object Adapter 
	Class adapter is not possible in java because can't extend multiple class 
*/

class AdapterPatternExample {

	public static void main(String args[]) { 

		AdapterPatternExample ape = new AdapterPatternExample();

		ape.new TextShapeAdapter(ape.new TextView()).BoundingBox();
	}

	interface Shape {
		void BoundingBox(); 
	}

	class Rectangle implements Shape {
		public void BoundingBox() {
			System.out.println("Simple bounding box....");
		}
	}

	class TextView {
		public void GetExtent() {
			System.out.println("TextView extent bound...");
		}
	}

	// adapter 
	class TextShapeAdapter implements Shape {
	
		// adaptee 
		private TextView tv;

		public TextShapeAdapter(TextView t) {
			this.tv = t;
		}

		public void BoundingBox() {
			// TextBox is not Shape, but acting like shape 
			System.out.println("Calling adaptee method from adapter...");
			tv.GetExtent();
		}
	}
}

File: Structural Pattern/Adapter/AdapterPatternExample.cpp
#include <iostream>
using namespace std;


class Shape 
{
public:
	virtual void Draw() = 0;
};

class TextView 
{
public: 
	TextView() 
	{
		//cout << "TextView..." << endl;
	}
	virtual void Show() 
	{
		cout << "Show the textview..." << endl;
	}
};


// class adapter 
class RectangleTextViewAdapter: public Shape, private TextView
{
public: 
	RectangleTextViewAdapter()
	{
		cout << "Rectangle-TextView Adapter..." << endl;
	}

	virtual void Draw() 
	{
		cout << "Drawing rectangle...." << endl;
		// call adaptee (textview) method 
		TextView::Show();
	}
};

int main() 
{
	Shape *s = new RectangleTextViewAdapter();
	s->Draw();


	return 0;
}

